packages/magic-link/src/accounts-magic-link.ts

```diff
@@ -0,0 +1,147 @@
+  /**
+   * Function called to verify the token.
+   */
+  verifyToken?: (token?: string, storedTokenEmail?: string) => boolean;
+  /**
+   * The number of milliseconds from when a link with a login token is sent until token expires and user can't login with it.
+   * Defaults to 15 minutes.
+   */
+  loginTokenExpiration?: number;
+}
+
+const defaultOptions = {
+  errors,
+  verifyToken,
+  // 15 minutes - 15 * 60 * 1000
+  loginTokenExpiration: 900000,
+};
+
+export default class AccountsMagicLink<CustomUser extends User = User>
+  implements AuthenticationService {
+  public serviceName = 'token';
+  public server!: AccountsServer;
+  private options: AccountsMagicLinkOptions & typeof defaultOptions;
+  private db!: DatabaseInterface<CustomUser>;
+
+  constructor(options: AccountsMagicLinkOptions = {}) {
+    this.options = { ...defaultOptions, ...options };
+  }
+
+  public setStore(store: DatabaseInterface<CustomUser>) {
+    this.db = store;
+  }
+
+  public async requestMagicLinkEmail(email: string): Promise<void> {
+    if (!email || !isString(email)) {
+      throw new AccountsJsError(
+        this.options.errors.invalidEmail,
+        RequestMagicLinkEmailErrors.InvalidEmail
+      );
+    }
+
+    const user = await this.db.findUserByEmail(email);
+    if (!user) {
+      throw new AccountsJsError(
+        this.options.errors.userNotFound,
+        RequestMagicLinkEmailErrors.UserNotFound
+      );
+    }
+
+    // Remove pre-existing login tokens on user
+    await this.db.removeAllLoginTokens(user.id);
+
+    const token = generateRandomToken();
+    await this.db.addLoginToken(user.id, email, token, 'magic-link-requested');
+
+    const requestMagicLinkMail = this.server.prepareMail(
+      email,
+      token,
+      this.server.sanitizeUser(user),
+      'token-login',
+      this.server.options.emailTemplates.magicLink,
+      this.server.options.emailTemplates.from
+    );
+
+    await this.server.options.sendMail(requestMagicLinkMail);
+  }
+
+  public async authenticate(params: LoginUserMagicLinkService): Promise<CustomUser> {
+    const { token } = params;
+    if (!token) {
+      throw new AccountsJsError(
+        this.options.errors.unrecognizedOptionsForLogin,
+        AuthenticateErrors.UnrecognizedOptionsForLogin
+      );
+    }
+    if (!isString(token)) {
+      throw new AccountsJsError(this.options.errors.matchFailed, AuthenticateErrors.MatchFailed);
+    }
+
+    const foundUser = await this.magicLinkAuthenticator(token);
+
+    // Remove all login tokens for user after login
+    await this.db.removeAllLoginTokens(foundUser.id);
+
+    return foundUser;
+  }
+
+  public isTokenExpired(tokenRecord: TokenRecord, expiryDate: number): boolean {
+    return Number(tokenRecord.when) + expiryDate < Date.now();
+  }
+
+  private async magicLinkAuthenticator(token: string): Promise<CustomUser> {
+    const foundUser: CustomUser | null = await this.db.findUserByLoginToken(token);
+
+    if (!foundUser) {
```

Comment:
if user is not found it means that token is invalid, better to throw an `incorrectToken` error

Reply:
also, can we rename `incorrectToken ` to `invalidToken` for consistency with password service?

Reply:
actually I was taking a look at the password service and we should throw `LoginTokenExpired `. That way, an attacker can't guess if the token is expired or does not exist.

Reply:
No ambigiousErrors handling? Always LoginTokenExpired? Would that also apply to the token record check itself?

Reply:
Yeah I think so, to be consistent with https://github.com/accounts-js/accounts/blob/master/packages/password/src/accounts-password.ts#L307